// @hirez_io/observer-spy/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.queueForAutoUnsubscribe = exports.autoUnsubscribe = exports.fakeTime = exports.subscribeSpyTo = exports.SubscriberSpy = exports.ObserverSpy = void 0;
/* istanbul ignore file */
var observer_spy_1 = require("./observer-spy");
Object.defineProperty(exports, "ObserverSpy", { enumerable: true, get: function () { return observer_spy_1.ObserverSpy; } });
var subscriber_spy_1 = require("./subscriber-spy");
Object.defineProperty(exports, "SubscriberSpy", { enumerable: true, get: function () { return subscriber_spy_1.SubscriberSpy; } });
var subscribe_spy_to_1 = require("./subscribe-spy-to");
Object.defineProperty(exports, "subscribeSpyTo", { enumerable: true, get: function () { return subscribe_spy_to_1.subscribeSpyTo; } });
var fake_time_1 = require("./fake-time");
Object.defineProperty(exports, "fakeTime", { enumerable: true, get: function () { return fake_time_1.fakeTime; } });
var auto_unsubscribe_1 = require("./auto-unsubscribe");
Object.defineProperty(exports, "autoUnsubscribe", { enumerable: true, get: function () { return auto_unsubscribe_1.autoUnsubscribe; } });
Object.defineProperty(exports, "queueForAutoUnsubscribe", { enumerable: true, get: function () { return auto_unsubscribe_1.queueForAutoUnsubscribe; } });
